cspann: cleanup and fixes for quantizers and search set

andy-kimball opened the pull request:
#### cspann: ensure unit vectors in quantizers when required

Construct a unit vector centroid in the quantizers when using the InnerProduct
or Cosine distance metric. This is required for accurate distances in the
Cosine case. In the InnerProduct case, it avoids partition centroids with high
magnitudes that "attract" vectors simply because of their magnitude.

In test builds, check that all query and data vectors are unit vectors when
using the Cosine distance metric.

Add more testing for Cosine and InnerProduct distance metrics, as used by the
quantizers.

#### cspann: remove centroid distances from quantized and search sets

Remove GetCentroidDistances() from QuantizedVectorSet and remove
CentroidDistance from SearchResult. The centroid distances were never
being used.

#### cspann: rename SearchResult.QuerySquaredDistance to QueryDistance

Now that we're introducing alternative distance metrics, the search result no
longer always returns a squared L2 distance. Rename the field to QueryDistance
to be agnostic to the distance metric that's used.

andy-kimball committed:
cspann: remove centroid distances from quantized and search sets

Remove GetCentroidDistances() from QuantizedVectorSet and remove
CentroidDistance from SearchResult. The centroid distances were never
being used.

Epic: CRDB-42943
Release note: None

andy-kimball committed:
cspann: rename SearchResult.QuerySquaredDistance to QueryDistance

Now that we're introducing alternative distance metrics, the search result no
longer always returns a squared L2 distance. Rename the field to QueryDistance
to be agnostic to the distance metric that's used.

Epic: CRDB-42943
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/146511)
<!-- Reviewable:end -->

andy-kimball committed:
cspann: ensure unit vectors in quantizers when required

Construct a unit vector centroid in the quantizers when using the InnerProduct
or Cosine distance metric. This is required for accurate distances in the
Cosine case. In the InnerProduct case, it avoids partition centroids with high
magnitudes that "attract" vectors simply because of their magnitude.

In test builds, check that all query and data vectors are unit vectors when
using the Cosine distance metric.

Add more testing for Cosine and InnerProduct distance metrics, as used by the
quantizers.

Epic: CRDB-42943
Release note: None

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 11 of 11 files at r1, 28 of 28 files at r2, 14 of 14 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146511#-:-OQ5Z39iCABD8nO5j3ti:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mw5h)

___
*[`pkg/sql/vecindex/cspann/quantize/rabitq.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146511#-OQ5XJLu97GJs-nf0xYt:-OQ5XJLu97GJs-nf0xYu:b-r7igs6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be381cafe71e83c729d8ac6d89841d957964dc1/pkg/sql/vecindex/cspann/quantize/rabitq.go#L363)):*
> ```Go
> 					distance = 0
> 				} else if distance > 2 {
> 					errorBound = min(errorBound - (distance - 2), 2)
> ```

Does this mean the error bound can end up negative? Does that matter?


<!-- Sent from Reviewable.io -->

andy-kimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/146511)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @DrewKimball and @mw5h)

___
*[`pkg/sql/vecindex/cspann/quantize/rabitq.go` line 363 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/146511#-OQ5XJLu97GJs-nf0xYt:-OQ6Lzxb2IyVfMd9q7qb:bocir2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8be381cafe71e83c729d8ac6d89841d957964dc1/pkg/sql/vecindex/cspann/quantize/rabitq.go#L363)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Does this mean the error bound can end up negative? Does that matter?
</blockquote></details>

hm, well, it's probably ok, but just to be safe i'll cap it


<!-- Sent from Reviewable.io -->

andy-kimball commented:
bors r=drewkimball
